### ISSUE TITLE
gazebo_ros_diff_drive crashes on delete

### ISSUE THREAD

**Issue by maxbader** [ros-simulation/gazebo_ros_pkgs#123]
Hi

It looks like that the gazebo_ros_diff_drive plug-in cases the gzserver to crash when a robot is removed (deleted) from the world

Ubuntu 12.04 (32bit)
ROS Hydro

Greetings

**Comment by scpeters** [comment]
Did you manage to get a backtrace?

**Comment by maxbader** [comment]
deleting the robot with the diff_drive plugin terminates the gzserver with the following msgs:

```
[ INFO] [1380650883.039866268, 0.047000000]: waitForService: Service [/gazebo/set_physics_properties] is now available.
[ INFO] [1380650883.119690601, 0.092000000]: Physics dynamic reconfigure ready.
[ INFO] [1380650893.319212552, 10.184000000]: Starting GazeboRosDiffDrive Plugin (ns = /pioneer2dx/)!
pure virtual method called
terminate called without an active exception
Aborted (core dumped)
[gazebo-2] process has died [pid 4626, exit code 134, cmd /opt/ros/hydro/lib/gazebo_ros/gzserver /home/max/projects/catkin_ws/src/v4r_gazebo/worlds/4walls.world __name:=gazebo __log:=/home/max/.ros/log/68978a4c-2ac4-11e3-a3f0-0024e8b0560b/gazebo-2.log].
log file: /home/max/.ros/log/68978a4c-2ac4-11e3-a3f0-0024e8b0560b/gazebo-2*.log
```

I am trying now to compile the diff_drive by my own and to add some debug outputs.
By the way the gazebo-2.log files do not exist

**Comment by maxbader** [comment]
Hi

I figured it out, the problem is within the destructor of plug-in.
The current code just deletes the rosnode_ poniter to without calling shutdown first.

```
GazeboRosDiffDrive::~GazeboRosDiffDrive() {
  delete rosnode_;
  delete transform_broadcaster_;
  }
```

This destructor solves the problem

```
GazeboRosDiffDrive::~GazeboRosDiffDrive() {
  delete transform_broadcaster_;  // I am not sure but I think it is better do delete the boradcaster first
  rosnode_->shutdown();
  delete rosnode_;
}
```

Greetings 
Markus

**Comment by j-rivero** [comment]
I haven't been able to reproduce this problem with ROS Jade. Please reopen if it still present on your system.

**Comment by Abhinav93** [comment]
Please reopen this the problem still persists in ROS meledic and gazebo 9. I think this is because of the empty destructor in differential_drive_controller
The destructor looks like tthis
`// Destructor
GazeboRosDiffDrive::~GazeboRosDiffDrive() {}`
I think ROS nodes should be shutdown as pointed out by @maxbader

**Comment by shiveshkhaitan** [comment]
Apparently this is supposed to be done by FiniChild.
https://github.com/ros-simulation/gazebo_ros_pkgs/blob/50a6382aa3f20962e6c1eedece20fc27178b2f9d/gazebo_plugins/src/gazebo_ros_diff_drive.cpp#L315-L322
But this function never gets called. A call to FiniChild() in the destructor  ```GazeboRosDiffDrive::~GazeboRosDiffDrive() { FiniChild (); }``` does exactly what @maxbader pointed out and it allows deletion of robot without any error..

**Comment by Abhinav93** [comment]
Hi,
Its been 20 days + and there has been no activity on the thread. Please look into the bug.

**Comment by chapulina** [comment]
Hi @Abhinav93 , you're welcome to open a pull request addressing the problem. We have no plans of working on it in the near future.

**Comment by chapulina** [comment]
Fixed by #1019 and #1021
